### PR TITLE
Add Transaction Simulation to the TXM

### PR DESCRIPTION
**Pull request by amit-momin**
- Add tx simulation to the TXM to allow product teams to test when a tx on ZK chains will fail due to out of counter (overflow) errors. Does not return any other types of errors. Non-zk chains should never error on this method.
- Enable tx simulation to easily adapt to chain specific methods
- Create a default case for all other chains

### CODE REVIEW

File: common/txmgr/txmgr.go [hunk 1]
```diff
@@ -603,6 +606,10 @@ func (b *Txm[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) CountTrans
 	return b.txStore.CountTransactionsByState(ctx, state, b.chainID)
 }
 
+func (b *Txm[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) CheckTxValidity(ctx context.Context, from ADDR, to ADDR, data []byte) error {
```
Comment (prashantkumar1982) [inline common/txmgr/txmgr.go:609]: This should also be added in the TxManager interface defined above in this file.

**Review by prashantkumar1982 — COMMENTED** [review]
Could you also update the txManager interface here if @FelixFan1992 agrees with this route?
https://github.com/smartcontractkit/chainlink/blob/develop/core/services/ocrcommon/transmitter.go#L18-L20

File: core/chains/evm/client/tx_simulator_test.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+	"github.com/smartcontractkit/chainlink/v2/core/internal/cltest"
+	"github.com/smartcontractkit/chainlink/v2/core/internal/testutils"
+)
+
+func TestSimulateTx_Default(t *testing.T) {
+	t.Parallel()
+
+	fromAddress := testutils.NewAddress()
+	toAddress := testutils.NewAddress()
+
```
Comment (DylanTinianov) [inline core/chains/evm/client/tx_simulator_test.go:23]: Nit: could define `ctx := testutils.Context(t)`

Comment (amit-momin) [reply core/chains/evm/client/tx_simulator_test.go:23]: Good call!

File: core/chains/evm/client/errors.go [hunk 3]
```diff
@@ -227,7 +228,11 @@ var zkSync = ClientErrors{
 	Fatal: regexp.MustCompile(`(?:: |^)(?:exceeds block gas limit|intrinsic gas too low|Not enough gas for transaction validation|Failed to pay the fee to the operator|Error function_selector = 0x, data = 0x|invalid sender. can't start a transaction from a non-account|max(?: priority)? fee per (?:gas|pubdata byte) higher than 2\^64-1|oversized data. max: \d+; actual: \d+)$`),
 }
 
-var clients = []ClientErrors{parity, geth, arbitrum, metis, substrate, avalanche, nethermind, harmony, besu, erigon, klaytn, celo, zkSync}
+var zkEvm = ClientErrors{
+	OutOfCounters: regexp.MustCompile(`(?:: |^)not enough keccak counters to continue the execution$`),
```
Comment (friedemannf) [inline core/chains/evm/client/errors.go:232]: This isn't the only error they return when out of counters, it depends on why they're being exceeded. There are currently eight ([source](https://github.com/0xPolygonHermez/zkevm-node/blob/aae30e9c79bdf363814e7fe2a3df9b34e855c998/state/runtime/runtime.go#L28-L43)):
```
not enough step counters to continue the execution
not enough keccak counters to continue the execution
not enough binary counters to continue the execution
not enough memory align counters to continue the execution
not enough arith counters to continue the execution
not enough padding counters to continue the execution
not enough poseidon counters to continue the execution
not enough sha256 counters to continue the execution
```

We could match them all using:
```suggestion
	OutOfCounters: regexp.MustCompile(`(?:: |^)not enough .* counters to continue the execution$`),
```

File: core/chains/evm/client/tx_simulator.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+	return nil
+}
+
+// eth_estimateGas returns out-of-counters (OOC) error if the transaction would result in an overflow
+func simulateTransactionDefault(ctx context.Context, client simulatorClient, msg ethereum.CallMsg) error {
+	var result hexutil.Big
+	errCall := client.CallContext(ctx, &result, "eth_estimateGas", toCallArg(msg), "pending")
+	jsonErr, _ := ExtractRPCError(errCall)
+	// Only return error if Zk OOC error is identified
+	if jsonErr != nil && jsonErr.Message == ErrOutOfCounters {
```
Comment (friedemannf) [inline core/chains/evm/client/tx_simulator.go:49]: This should also check against all eight possible errors

File: core/chains/evm/client/tx_simulator.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+			GasUsed              string
+			UsedKeccakHashes     string
+			UsedPoseidonHashes   string
+			UsedPoseidonPaddings string
+			UsedMemAligns        string
+			UsedArithmetics      string
+			UsedBinaries         string
+			UsedSteps            string
+			UsedSHA256Hashes     string
+		}
```
Comment (friedemannf) [inline core/chains/evm/client/tx_simulator.go:68]: We could also check if the `CountersUsed` exceed the `CountersLimit` in the response? This way wouldn't be relying on the error message, which might be preferrable?

**Review by friedemannf — CHANGES_REQUESTED** [review]

File: core/chains/evm/client/errors.go (hunk 3 above)
Comment (amit-momin) [reply core/chains/evm/client/errors.go]: Ah didn't know that, thanks for calling it out. I've updated the regex accordingly

File: core/chains/evm/client/tx_simulator.go (hunk 4 above)
Comment (amit-momin) [reply core/chains/evm/client/tx_simulator.go]: Realized we filter for OOC errors using `ClassifySendError` in `SimulateTransaction` anyways so this check actually isn't needed. I've removed it in the latest commit

File: core/chains/evm/client/tx_simulator.go (hunk 5 above)
Comment (amit-momin) [reply core/chains/evm/client/tx_simulator.go]: Good point! I've added some validation checking these fields against each other. Since we're targeting OOC errors only though, I excluded the validation on `GasUsed`.

File: core/chains/evm/client/tx_simulator.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+// Helper method for zkEvm to determine if response indicates an overflow
+func detectZkEvmCounterOverflow(result zkEvmEstimateCountResponse) bool {
+	if result.CountersUsed.UsedKeccakHashes > result.CountersLimit.MaxKeccakHashes ||
+		result.CountersUsed.UsedPoseidonHashes > result.CountersLimit.MaxPoseidonHashes ||
+		result.CountersUsed.UsedPoseidonPaddings > result.CountersLimit.MaxPoseidonPaddings ||
+		result.CountersUsed.UsedMemAligns > result.CountersLimit.MaxMemAligns ||
+		result.CountersUsed.UsedArithmetics > result.CountersLimit.MaxArithmetics ||
+		result.CountersUsed.UsedBinaries > result.CountersLimit.MaxBinaries ||
+		result.CountersUsed.UsedSteps > result.CountersLimit.MaxSteps ||
+		result.CountersUsed.UsedSHA256Hashes > result.CountersLimit.MaxSHA256Hashes {
```
Comment (friedemannf) [inline core/chains/evm/client/tx_simulator.go:104]: Comparing these hex values lexicographically will not work as `"0x9">"0x123" == true`. Setting the fields to e.g. `hexutil.Uint64` should unmarshal them correctly.
Tests should have caught that, as your input contains that case but the `&&` in `simulateTransactionZkEvm ` masked it.

File: core/chains/evm/client/tx_simulator.go [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+// zkEVM implemented a custom zkevm_estimateCounters method to detect if a transaction would result in an out-of-counters (OOC) error
+func simulateTransactionZkEvm(ctx context.Context, client simulatorClient, lggr logger.SugaredLogger, msg ethereum.CallMsg) error {
+	var result zkEvmEstimateCountResponse
+	err := client.CallContext(ctx, &result, "zkevm_estimateCounters", toCallArg(msg), "pending")
+	if err != nil {
+		return fmt.Errorf("failed to simulate tx: %w", err)
+	}
+	if detectZkEvmCounterOverflow(result) && len(result.OocError) > 0 {
```
Comment (friedemannf) [inline core/chains/evm/client/tx_simulator.go:88]: The `&&` condition here seems redundant, might be better to err on the side of caution and make it a `||`?
The error message might be changed in the future and not match the regex and this way we could still detect it.

**Review by friedemannf — CHANGES_REQUESTED** [review]

File: core/chains/evm/client/tx_simulator.go (hunk 7 above)
Comment (amit-momin) [reply core/chains/evm/client/tx_simulator.go:88]: I realized I may have misunderstood the original [comment](https://github.com/smartcontractkit/chainlink/pull/12503#discussion_r1536223363) that prompted me to add `detectZkEvmCounterOverflow`. I don't think we want to completely break reliance on the error message. We want to pass it through our `ClassifySendError` [method](https://github.com/smartcontractkit/chainlink/pull/12503/files#diff-ed85d15901e7cfbfdfa625b140d903c8ec415cf09c24373330e695b874b3053aR36). To your point if they change the message and our regex no longer matches it, you're right we'd stop detecting the overflow but we have some work in progress to allow regex's to be configurable through toml which would allow us to quickly address this issue in the future. If we solely relied on the used vs max comparison from the response, we'd maybe run the risk of a new counter getting added that we don't account for. Don't know how likely that'd be but the only path forward there would be a code change limiting how quickly we could address it.

Comment (amit-momin) [reply core/chains/evm/client/tx_simulator.go:88]: Since we're passing this through our `ClassifySendError` method, I think it's ok if we just pass the error message through if one exists without checking further. To avoid the risk of counter fields changing, I'm feeling inclined to remove `detectZkEvmCounterOverflow`.

File: core/chains/evm/client/tx_simulator.go (hunk 6 above)
Comment (amit-momin) [reply core/chains/evm/client/tx_simulator.go:104]: Shoot good catch! I changed around the response types a few times during dev and forgot these aren't already uint64 😅. But I'll wait to address this once we decide whether we want this method or not in the other comment.

File: core/chains/evm/client/tx_simulator.go [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+func SimulateTransaction(ctx context.Context, client simulatorClient, lggr logger.SugaredLogger, chainType config.ChainType, msg ethereum.CallMsg) error {
+	var err error
+	switch chainType {
+	case config.ChainZkEvm:
+		err = simulateTransactionZkEvm(ctx, client, lggr, msg)
+	default:
+		err = simulateTransactionDefault(ctx, client, msg)
+	}
+	// ClassifySendError will not have the proper fields for logging within the method due to the empty Transaction passed
+	code := ClassifySendError(err, lggr, &types.Transaction{}, msg.From, chainType.IsL2())
```
[the remark this answers is not in the document]
Comment (amit-momin) [reply core/chains/evm/client/tx_simulator.go:36]: I think it was a suggestion from Dimitris so we could take advantage of configurable errors after Nick's changes unless I misunderstood.  You bring up a good point though that this would be mixing different types of errors though. I could setup a different parser for tx simulation errors but would that be overkill?

File: common/config/chaintype.go [hunk 9]
```diff
@@ -19,6 +19,7 @@ const (
 	ChainScroll          ChainType = "scroll"
 	ChainWeMix           ChainType = "wemix"
 	ChainXDai            ChainType = "xdai" // Deprecated: use ChainGnosis instead
+	ChainZkEvm           ChainType = "zkevm"
```
Comment (prashantkumar1982) [inline common/config/chaintype.go:22]: Why is this needed?
Didn't see any behavior depending on this type.

Comment (amit-momin) [reply common/config/chaintype.go:22]: Oh good call, missed cleaning this out after removing the zkevm specific method

File: core/chains/evm/client/chain_client.go [hunk 10]
```diff
@@ -269,3 +270,12 @@ func (c *chainClient) TransactionReceipt(ctx context.Context, txHash common.Hash
 func (c *chainClient) LatestFinalizedBlock(ctx context.Context) (*evmtypes.Head, error) {
 	return c.multiNode.LatestFinalizedBlock(ctx)
 }
+
+func (c *chainClient) CheckTxValidity(ctx context.Context, from common.Address, to common.Address, data []byte) error {
```
Comment (prashantkumar1982) [inline core/chains/evm/client/chain_client.go:274]: How should the product understand if the received error is OOC type?

Comment (amit-momin) [reply core/chains/evm/client/chain_client.go:274]: Since we standardize the error we send back, they could use [ErrOutOfCounters](https://github.com/smartcontractkit/chainlink/pull/12503/files#diff-ed85d15901e7cfbfdfa625b140d903c8ec415cf09c24373330e695b874b3053aR14) to determine if the error is an OOC error. I think to make it simpler for them though we'd maybe need to return a second output either another error or a flag. But I took this approach in efforts to keep the return a single field.

**Review by prashantkumar1982 — DISMISSED** [review]
Looks good for now to unblock ZK chains.

For future, I think we need to come up with a proper error handling strategy for the Chain Client. All RPC calls in the Chain Client should support a standard error parsing logic, and categorization.

File: core/chains/evm/client/chain_client.go [hunk 11]
```diff
@@ -269,3 +270,12 @@ func (c *chainClient) TransactionReceipt(ctx context.Context, txHash common.Hash
 func (c *chainClient) LatestFinalizedBlock(ctx context.Context) (*evmtypes.Head, error) {
 	return c.multiNode.LatestFinalizedBlock(ctx)
 }
+
+func (c *chainClient) CheckTxValidity(ctx context.Context, from common.Address, to common.Address, data []byte) *SendError {
```
Comment (dimriou) [inline core/chains/evm/client/chain_client.go:274]: nit: would it make sense to rename this function to something like `CheckTxOverflow` or something like that? `CheckTxValidity` seems a bit vague.

File: core/chains/evm/client/tx_simulator.go
Comment (dimriou) [inline core/chains/evm/client/tx_simulator.go:1]: Why create a separate simulatorClient instead of adding the logic directly to the existing client?

Comment (amit-momin) [reply core/chains/evm/client/tx_simulator.go:1]: Before when we were using the custom zkEvm method, it was so we could maintain chain specific code without crowding the client. It's possible now that it's simplified but I think there's a realistic possibility that we may need to support a custom method for a chain that's still in-progress.

File: core/chains/evm/client/chain_client.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (amit-momin) [reply core/chains/evm/client/chain_client.go:274]: Prashant and I talked this over a bit in the thread. I don't think this could be handled with just a bool or a basic error type. There might be some additional errors that the product teams could check for when simulating like service unavailable or timeout. If it was a simple bool or error type, they wouldn't be able to differentiate between these which is why we opted to return `SendError`. That way they could check the specific types of errors with the built in methods like `IsOutOfCounters()`

Comment (amit-momin) [reply core/chains/evm/client/chain_client.go:274]: I agree with this change

File: core/chains/evm/client/tx_simulator.go
Comment (dimriou) [reply core/chains/evm/client/tx_simulator.go:1]: I might regret this in the future, but I still prefer to have this inside `CheckTxOverflow` method. Right now `CheckTxOverflow` is like a passthrough and has no logic. Chain client is supposed to serve exactly this purpose, which is to implement chain-specific logic that can't go inside multinode. As a reader, it would be easier for me to just read the entire thing under one file instead of having to understand why `tx_simulator` was created. I don't have a hard stance on this, so if you guys think it's cleaner as it is, then I'm ok keeping it this way.

Comment (amit-momin) [reply core/chains/evm/client/tx_simulator.go:1]: I would maybe argue that chain client is more so EVM specific code. So if something requires chain specific code where different EVM chains have different implementation, my opinion is we'd want to separate that from the chain client. That being said, we skipped on using the zkevm chain specific code here so I'm not against moving this to the chain client. We could always separate this out again later when there's a need. But before I do, let me get agreement from others.

File: core/chains/evm/client/chain_client.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (prashantkumar1982) [reply core/chains/evm/client/chain_client.go:274]: There's an opportunity to clean up our error responses handling across ALL of chain client interface.
Can't be tackled in this PR, but likely a bigger effort into how to handle errors in standard way, for all of the Chain Client interfaces. Maybe a new ticket.
Created one here: https://smartcontract-it.atlassian.net/browse/BCI-2863

File: core/chains/evm/client/tx_simulator.go
Comment (prashantkumar1982) [reply core/chains/evm/client/tx_simulator.go:1]: I would be ok with anything here.
Either make some changes here, or wait for the final solution to be clearer wrt Linea, Scroll and other ZK chains, and then make the appropriate code changes.

**Review by prashantkumar1982 — APPROVED** [review]

File: core/chains/evm/client/tx_simulator.go
Comment (amit-momin) [reply core/chains/evm/client/tx_simulator.go:1]: I'd prefer to wait to avoid another back and forth while product teams are waiting. But I agree we'll have more changes come to this code in the near future when we can make this decision.
